api/sendfild/service/service.go

```diff
@@ -0,0 +1,713 @@
+			}
+		}
+	}()
+}
+
+func (s *Service) SendFil(ctx context.Context, req *pb.SendFilRequest) (*pb.SendFilResponse, error) {
+	f, err := address.NewFromString(req.From)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing from address: %v", err)
+	}
+	t, err := address.NewFromString(req.To)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "parsing to address: %v", err)
+	}
+	nanoAmount := (&big.Int{}).SetInt64(req.AmountNanoFil)
+	factor := (&big.Int{}).SetInt64(int64(math.Pow10(9)))
+	amount := (&big.Int{}).Mul(nanoAmount, factor)
+	msg := &types.Message{
+		From:  f,
+		To:    t,
+		Value: types.BigInt{Int: amount},
+	}
+	client, cls, err := s.clientBuilder(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "creating filecoin client: %v", err)
+	}
+	defer cls()
+
+	sm, err := client.MpoolPushMessage(ctx, msg, nil)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "pushing message: %v", err)
+	}
+
+	now := time.Now()
+
+	tx := txn{
+		ID:            primitive.NewObjectID(),
+		From:          req.From,
+		To:            req.To,
+		AmountNanoFil: req.AmountNanoFil,
+		MessageCids:   []msgCid{{Cid: sm.Message.Cid().String(), CreatedAt: now}},
+		MessageState:  pb.MessageState_MESSAGE_STATE_PENDING,
+		CreatedAt:     now,
+		UpdatedAt:     now,
+	}
+
+	if _, err = s.col.InsertOne(ctx, &tx); err != nil {
+		return nil, status.Errorf(codes.Internal, "inserting txn into collection: %v", err)
+	}
+
+	wait := s.wait(tx)
+
+	if req.Wait {
+		res := <-wait
+		if res.err != nil {
+			return nil, status.Errorf(codes.Internal, "waiting for result: %v", res.err)
+		}
+		tx = res.txn
+	}
+
+	pbTx, err := toPbTxn(tx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "converting to pb txn: %v", err)
+	}
+
+	return &pb.SendFilResponse{Txn: pbTx}, nil
+}
+
+func (s *Service) Txn(ctx context.Context, req *pb.TxnRequest) (*pb.TxnResponse, error) {
+	res := s.col.FindOne(ctx, bson.M{"message_cids.cid": req.MessageCid})
+	if res.Err() == mongo.ErrNoDocuments {
+		return nil, status.Error(codes.NotFound, "no txn found for cid")
+	}
+	if res.Err() != nil {
+		return nil, status.Errorf(codes.Internal, "querying for cid txn: %v", res.Err())
+	}
+
+	var tx txn
+	if err := res.Decode(&tx); err != nil {
+		return nil, status.Errorf(codes.Internal, "decoding cid txn result: %v", res.Err())
+	}
+
+	if tx.MessageState != pb.MessageState_MESSAGE_STATE_ACTIVE && tx.MessageState != pb.MessageState_MESSAGE_STATE_FAILED && req.Wait {
+		res := <-s.wait(tx)
+		if res.err != nil {
+			return nil, status.Errorf(codes.Internal, "waiting for result: %v", res.err)
+		}
+		tx = res.txn
+	}
+
+	pbTx, err := toPbTxn(tx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "converting to pb txn: %v", err)
+	}
+
+	return &pb.TxnResponse{Txn: pbTx}, nil
+}
+
+func (s *Service) ListTxns(ctx context.Context, req *pb.ListTxnsRequest) (*pb.ListTxnsResponse, error) {
```

Comment:
Our "list" api. I went pretty deep with the query filter support, so all the below code is a bit hard to read, but it just maps request params to find options and filters. This is all tested pretty well.

Reply:
Some combinations of query options are not valid, so this code just parses valid combinations and invalid combinations are no-op. The valid combinations are common sense kinda stuff, not idea feelback to the caller, but hopefully fine for an internal service like this.